x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/Querier.java

```diff
@@ -535,14 +536,15 @@ protected List<BucketExtractor> initBucketExtractors(SearchResponse response) {
             List<QueryContainer.FieldInfo> refs = query.fields();
 
             List<BucketExtractor> exts = new ArrayList<>(refs.size());
-            ConstantExtractor totalCount = new ConstantExtractor(response.getHits().getTotalHits().value);
+            TotalHits totalHits = response.getHits().getTotalHits();
+            ConstantExtractor totalCount = totalHits == null ? null : new ConstantExtractor(totalHits.value);
```

Comment:
This is trappy behavior since now the null can leak.
Just extend ConstantExtractor (or BucketExtractor) with a dedicated TotalHits implementation that handles the null case and just pas totalHits object to it and handle it there by getting the concrete value or -1.
This would have minimal changes to the code.

Reply:
> looking at the code we only enable total hit counts when needed so by default it should be disabled.

The Java API is misleading here, `track_total_hits` is actually a number, by default it's 10.000; I'm also curious to see the impact on the benchmarks.

> Just extend ConstantExtractor (or BucketExtractor) with a dedicated TotalHits implementation that handles the null case and just pas totalHits object to it and handle it there by getting the concrete value or -1.

Do we really need to extend `ConstantExtractor` for this? If we just want to return a -1, we can just pass it to `ConstantExtractor` itself.

Reply:
My understanding is that this -1 should never be returned (and that's why Andrei's exception throwing proposal [above](https://github.com/elastic/elasticsearch/pull/89106#discussion_r941201104), I assume), since the `GlobalCountRef.INSTANCE` ref is only used when accurate counting is required.
Don't know if we should go this way, but if we do add a TotalHitsExtractor, it might make sense to keep throwing an exception the exception on requiring the count when not actually needed.
If we don't go that way, I'd suggest replacing the `-1` with `org.elasticsearch.search.internal.SearchContext.TRACK_TOTAL_HITS_DISABLED`.

Reply:
Any value we pass here, will always be a wrong value. 
About `org.elasticsearch.search.internal.SearchContext.TRACK_TOTAL_HITS_DISABLED`, that's the config value to disable the total hits, while here we are returning a calculated value, so IMHO they do not really represent the same thing.
My initial intention was to keep failing if something is wrong in the planner (so the Holder, that delayed the calculation of the totalHits at the moment when it was actually required by the execution plan). This could save us from problems introduced by future changes; on the other hand, passing a default value (eg. -1) could hide the consequences of a wrong planning. Anyway, for current implementation I don't see a possible path where this could really happen.

I'd say let's just reach an agreement, the impact of this choice should not be dramatic.